File: client-python-urllib3/src/epistola_client/auth/jwt_signer.py
"""Creates and signs short-lived JWTs for self-signed JWT authentication with Epistola.

Each token contains:

* ``iss``: the consumer ID
* ``iat``: issued-at timestamp
* ``exp``: expiry (iat + token_lifetime)
* ``jti``: unique nonce (UUID) for replay protection

Example::

    signer = (
        JwtSigner.builder()
        .consumer_id("invoice-service")
        .private_key(JwtSigner.load_private_key("private.pem"))
        .build()
    )
"""

from __future__ import annotations

import datetime
import uuid
from typing import Optional

import jwt
from cryptography.hazmat.primitives.asymmetric import ec, rsa
from cryptography.hazmat.primitives.serialization import load_pem_private_key

_DEFAULT_LIFETIME = datetime.timedelta(seconds=60)


class JwtSigner:
    """Immutable signer that mints freshly signed JWTs for the configured consumer."""

    def __init__(self, consumer_id: str, private_key, algorithm: str, token_lifetime: datetime.timedelta) -> None:
        self._consumer_id = consumer_id
        self._private_key = private_key
        self._algorithm = algorithm
        self._token_lifetime = token_lifetime

    @staticmethod
    def builder() -> "JwtSignerBuilder":
        """Create a new :class:`JwtSignerBuilder`."""
        return JwtSignerBuilder()

    @staticmethod
    def load_private_key(path: str):
        """Load a private key from a PEM file (RSA or EC P-256, PKCS#8 ``BEGIN PRIVATE KEY``)."""
        with open(path, "rb") as handle:
            return JwtSigner.parse_private_key_pem(handle.read())

    @staticmethod
    def parse_private_key_pem(pem):
        """Parse a PEM-encoded private key (RSA or EC P-256, PKCS#8 ``BEGIN PRIVATE KEY``)."""
        if isinstance(pem, str):
            pem = pem.encode("utf-8")
        try:
            return load_pem_private_key(pem, password=None)
        except Exception as exc:  # noqa: BLE001 - normalise to a clear ValueError
            raise ValueError(
                "Failed to parse private key. Supported formats: RSA, EC (P-256) in PKCS#8 PEM format."
            ) from exc

    def create_token(self) -> str:
        """Create a freshly signed JWT with a new ``iat``, ``exp``, and ``jti``."""
        now = datetime.datetime.now(datetime.timezone.utc)
        payload = {
            "iss": self._consumer_id,
            "iat": int(now.timestamp()),
            "exp": int((now + self._token_lifetime).timestamp()),
            "jti": str(uuid.uuid4()),
        }
        return jwt.encode(payload, self._private_key, algorithm=self._algorithm)


class JwtSignerBuilder:
    """Fluent builder for :class:`JwtSigner`."""

    def __init__(self) -> None:
        self._consumer_id: Optional[str] = None
        self._private_key = None
        self._token_lifetime = _DEFAULT_LIFETIME

    def consumer_id(self, consumer_id: str) -> "JwtSignerBuilder":
        """Set the consumer ID used as the JWT ``iss`` claim."""
        if not consumer_id or not consumer_id.strip():
            raise ValueError("consumer_id must not be blank")
        self._consumer_id = consumer_id
        return self

    def private_key(self, private_key) -> "JwtSignerBuilder":
        """Set the private key used to sign tokens (from :meth:`JwtSigner.load_private_key` /
        :meth:`JwtSigner.parse_private_key_pem`).
        """
        self._private_key = private_key
        return self

    def token_lifetime(self, lifetime: datetime.timedelta) -> "JwtSignerBuilder":
        """Set the token lifetime (default: 60 seconds)."""
        if lifetime <= datetime.timedelta(0):
            raise ValueError("token_lifetime must be positive")
        self._token_lifetime = lifetime
        return self

    def build(self) -> JwtSigner:
        """Build the immutable :class:`JwtSigner`."""
        if self._consumer_id is None:
            raise ValueError("consumer_id is required")
        if self._private_key is None:
            raise ValueError("private_key is required")

        if isinstance(self._private_key, rsa.RSAPrivateKey):
            algorithm = "RS256"
        elif isinstance(self._private_key, ec.EllipticCurvePrivateKey):
            algorithm = "ES256"
        else:
            raise ValueError(
                f"Unsupported key type: {type(self._private_key).__name__}. Supported: RSA (2048+), EC (P-256)"
            )

        return JwtSigner(self._consumer_id, self._private_key, algorithm, self._token_lifetime)

File: contracts/api/clients/dotnet-httpclient/src/Epistola.Client/Auth/JwtSigner.cs
// SPDX-FileCopyrightText: Epistola Nederland B.V.
//
// SPDX-License-Identifier: EUPL-1.2

using System;
using System.IdentityModel.Tokens.Jwt;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Security.Claims;
using System.Security.Cryptography;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.IdentityModel.Tokens;

namespace Epistola.Client.Auth;

/// <summary>
/// Creates and signs short-lived JWTs for self-signed JWT authentication with Epistola.
///
/// Each token contains:
/// <list type="bullet">
/// <item><c>iss</c>: the consumer ID</item>
/// <item><c>iat</c>: issued-at timestamp</item>
/// <item><c>exp</c>: expiry (iat + tokenLifetime)</item>
/// <item><c>jti</c>: unique nonce (GUID) for replay protection</item>
/// </list>
///
/// <code>
/// var signer = JwtSigner.Builder()
///     .ConsumerId("invoice-service")
///     .PrivateKey(JwtSigner.LoadPrivateKey("private.pem"))
///     .Build();
/// </code>
/// </summary>
public sealed class JwtSigner
{
    private readonly string _consumerId;
    private readonly SigningCredentials _credentials;
    private readonly TimeSpan _tokenLifetime;

    private JwtSigner(string consumerId, SigningCredentials credentials, TimeSpan tokenLifetime)
    {
        _consumerId = consumerId;
        _credentials = credentials;
        _tokenLifetime = tokenLifetime;
    }

    /// <summary>Creates a new <see cref="JwtSignerBuilder"/>.</summary>
    public static JwtSignerBuilder Builder() => new();

    /// <summary>Loads a private key from a PEM file (RSA or EC P-256, PKCS#8 <c>BEGIN PRIVATE KEY</c>).</summary>
    public static AsymmetricAlgorithm LoadPrivateKey(string path) => ParsePrivateKeyPem(File.ReadAllText(path));

    /// <summary>Parses a PEM-encoded private key (RSA or EC P-256, PKCS#8 <c>BEGIN PRIVATE KEY</c>).</summary>
    public static AsymmetricAlgorithm ParsePrivateKeyPem(string pem)
    {
        try
        {
            var rsa = RSA.Create();
            rsa.ImportFromPem(pem);
            return rsa;
        }
        catch (Exception)
        {
            try
            {
                var ecdsa = ECDsa.Create();
                ecdsa.ImportFromPem(pem);
                return ecdsa;
            }
            catch (Exception)
            {
                throw new ArgumentException(
                    "Failed to parse private key. Supported formats: RSA, EC (P-256) in PKCS#8 PEM format.");
            }
        }
    }

    /// <summary>Creates a freshly signed JWT with a new <c>iat</c>, <c>exp</c>, and <c>jti</c>.</summary>
    public string CreateToken()
    {
        var now = DateTime.UtcNow;
        var handler = new JwtSecurityTokenHandler();
        var token = new JwtSecurityToken(
            issuer: _consumerId,
            audience: null,
            claims: new[]
            {
                new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString()),
                new Claim(JwtRegisteredClaimNames.Iat,
                    new DateTimeOffset(now).ToUnixTimeSeconds().ToString(),
                    ClaimValueTypes.Integer64),
            },
            notBefore: now,
            expires: now.Add(_tokenLifetime),
            signingCredentials: _credentials);
        return handler.WriteToken(token);
    }

    /// <summary>
    /// Creates a <see cref="DelegatingHandler"/> that sets <c>Authorization: Bearer &lt;jwt&gt;</c> on
    /// every request, minting a fresh token each time.
    /// </summary>
    public DelegatingHandler Handler() => new JwtSignerHandler(this);

    /// <summary>Fluent builder for <see cref="JwtSigner"/>.</summary>
    public sealed class JwtSignerBuilder
    {
        private string? _consumerId;
        private AsymmetricAlgorithm? _privateKey;
        private TimeSpan _tokenLifetime = TimeSpan.FromSeconds(60);

        /// <summary>Sets the consumer ID used as the JWT <c>iss</c> claim.</summary>
        public JwtSignerBuilder ConsumerId(string consumerId)
        {
            if (string.IsNullOrWhiteSpace(consumerId)) throw new ArgumentException("consumerId must not be blank", nameof(consumerId));
            _consumerId = consumerId;
            return this;
        }

        /// <summary>Sets the private key used to sign tokens (from <see cref="LoadPrivateKey"/> / <see cref="ParsePrivateKeyPem"/>).</summary>
        public JwtSignerBuilder PrivateKey(AsymmetricAlgorithm privateKey)
        {
            _privateKey = privateKey;
            return this;
        }

        /// <summary>Sets the token lifetime (default: 60 seconds).</summary>
        public JwtSignerBuilder TokenLifetime(TimeSpan lifetime)
        {
            if (lifetime <= TimeSpan.Zero) throw new ArgumentException("tokenLifetime must be positive", nameof(lifetime));
            _tokenLifetime = lifetime;
            return this;
        }

        /// <summary>Builds the immutable <see cref="JwtSigner"/>.</summary>
        public JwtSigner Build()
        {
            var id = _consumerId ?? throw new InvalidOperationException("consumerId is required");
            var key = _privateKey ?? throw new InvalidOperationException("privateKey is required");
            var credentials = key switch
            {
                RSA rsa => new SigningCredentials(new RsaSecurityKey(rsa), SecurityAlgorithms.RsaSha256),
                ECDsa ecdsa => new SigningCredentials(new ECDsaSecurityKey(ecdsa), SecurityAlgorithms.EcdsaSha256),
                _ => throw new ArgumentException(
                    $"Unsupported key type: {key.GetType().Name}. Supported: RSA (2048+), EC (P-256)"),
            };
            return new JwtSigner(id, credentials, _tokenLifetime);
        }
    }

    private sealed class JwtSignerHandler : DelegatingHandler
    {
        private readonly JwtSigner _signer;

        public JwtSignerHandler(JwtSigner signer)
        {
            _signer = signer;
        }

        protected override Task<HttpResponseMessage> SendAsync(
            HttpRequestMessage request,
            CancellationToken cancellationToken)
        {
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _signer.CreateToken());
            return base.SendAsync(request, cancellationToken);
        }
    }
}
